### ISSUE TITLE
JSON columns dont appear to be working

### ISSUE THREAD

**Issue by matthewjumpsoffbuildings** [colopl/laravel-spanner#149]
I have a model with a json column. As usual I have `$casts = ['json_column' => 'array']` in the model definition, but for some reason every time I try to save/update the model I get this error

```
Value of type STRING cannot be assigned to data, which has type JSON 
```

**Comment by taka-oyama** [comment]
There is no good solution to this. 
The query builder has no way of distinguishing between a normal string and a json string.

**Comment by taka-oyama** [comment]
Spanner didn't have JSON types initially (2019) and we just stored JSON as STRING(MAX).
That was good enough for our use case.

I'll update the docs to say it's not supported for the moment.

**Comment by taka-oyama** [comment]
PR is always welcome btw.

**Comment by matthewjumpsoffbuildings** [comment]
I haven't dealt with custom database/eloquent drivers before so its hard to follow the chain of how data is set/escaped before being passed to the `execute` statement.

I am trying to work out how the `escape` functions work here:
https://github.com/colopl/laravel-spanner/blob/724159abe1ee844fd4cec7d16c08c50d4d200d7e/src/Connection.php#L419C3-L442C6

I was thinking perhaps if you pass an associative array, the `escapeArray()` method, instead of throwing an exception, could treat it as JSON? 

But in my experimenting it doesnt even seem that the `escapeArray()` method gets called, at least not before this
https://github.com/colopl/laravel-spanner/blob/724159abe1ee844fd4cec7d16c08c50d4d200d7e/src/Connection.php#L319

**Comment by taka-oyama** [comment]
> I am trying to work out how the escape functions work here

From what I can recall, `escape` is only used for the `Builder::toRawSql` which is only really used for logging and debugging.

**Comment by taka-oyama** [comment]
> I was thinking perhaps if you pass an associative array, the escapeArray() method, instead of throwing an exception, could treat it as JSON?

That could potentially work but you might need to pass it as an object since empty arrays can't be treated as assoc array.

**Comment by matthewjumpsoffbuildings** [comment]
The solution is a custom Cast with a custom ValueInterface, as per this comment

https://github.com/colopl/laravel-spanner/pull/154#issuecomment-1849330874

Should add this to the docs

**Comment by matthewjumpsoffbuildings** [comment]
PR #265 sorts this
